File: ros/catkin_ws/src/simple_robotic_arm/scripts/yolo_obstacle_detector.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""ROS Image -> YOLO ONNX obstacle boxes -> ROS debug Image.

The model is a one-class obstacle detector exported from Ultralytics YOLO.
Inference uses onnxruntime so the node works with ROS Noetic system Python and
old OpenCV 4.2; OpenCV is only used for resize, NMS, and drawing.
"""
from __future__ import print_function

import os
import sys
import time

import cv2
import numpy as np
import onnxruntime as ort
import rospy
from cv_bridge import CvBridge, CvBridgeError
from sensor_msgs.msg import Image


def letterbox(bgr, size):
    h, w = bgr.shape[:2]
    scale = min(float(size) / float(w), float(size) / float(h))
    nw, nh = int(round(w * scale)), int(round(h * scale))
    resized = cv2.resize(bgr, (nw, nh), interpolation=cv2.INTER_LINEAR)
    canvas = np.full((size, size, 3), 114, dtype=np.uint8)
    pad_x = (size - nw) // 2
    pad_y = (size - nh) // 2
    canvas[pad_y:pad_y + nh, pad_x:pad_x + nw] = resized
    return canvas, scale, pad_x, pad_y


def clamp(v, lo, hi):
    return max(lo, min(hi, v))


class YoloObstacleDetector(object):
    def __init__(self):
        self.bridge = CvBridge()
        pkg_dir = os.path.abspath(os.path.join(os.path.dirname(__file__), ".."))
        default_model = os.path.join(pkg_dir, "models", "obstacle_yolo11n.onnx")

        self.image_topic = rospy.get_param("~image_topic", "/usb_cam/image_raw")
        self.debug_topic = rospy.get_param("~debug_image_topic", "/obstacle_detector/debug")
        self.model_path = rospy.get_param("~model_path", default_model)
        self.input_size = int(rospy.get_param("~input_size", 512))
        self.conf_thresh = float(rospy.get_param("~conf_thresh", 0.05))
        self.nms_thresh = float(rospy.get_param("~nms_thresh", 0.45))
        self.max_det = int(rospy.get_param("~max_det", 30))
        self.process_every_n = max(1, int(rospy.get_param("~process_every_n", 1)))
        self.output_width = int(rospy.get_param("~output_width", 320))
        self.draw_fps = bool(rospy.get_param("~draw_fps", True))

        if not os.path.exists(self.model_path):
            raise RuntimeError("model not found: %s" % self.model_path)

        self.session = ort.InferenceSession(self.model_path, providers=["CPUExecutionProvider"])
        self.input_name = self.session.get_inputs()[0].name
        self.pub = rospy.Publisher(self.debug_topic, Image, queue_size=1)
        self.sub = rospy.Subscriber(self.image_topic, Image, self.on_image, queue_size=1, buff_size=2 ** 24)
        self.frame_idx = 0
        self.last_boxes = []
        self.last_fps = 0.0
        rospy.loginfo("yolo_obstacle_detector: image=%s debug=%s model=%s conf=%.3f", self.image_topic, self.debug_topic, self.model_path, self.conf_thresh)

    def preprocess(self, bgr):
        padded, scale, pad_x, pad_y = letterbox(bgr, self.input_size)
        rgb = cv2.cvtColor(padded, cv2.COLOR_BGR2RGB)
        blob = rgb.astype(np.float32) / 255.0
        blob = np.transpose(blob, (2, 0, 1))[None, :, :, :]
        return blob, scale, pad_x, pad_y

    def postprocess(self, output, shape, scale, pad_x, pad_y):
        h, w = shape[:2]
        pred = output[0]
        if pred.shape[0] == 5:
            pred = pred.T
        boxes = []
        scores = []
        for row in pred:
            conf = float(row[4])
            if conf < self.conf_thresh:
                continue
            cx, cy, bw, bh = [float(x) for x in row[:4]]
            x1 = (cx - bw / 2.0 - pad_x) / scale
            y1 = (cy - bh / 2.0 - pad_y) / scale
            x2 = (cx + bw / 2.0 - pad_x) / scale
            y2 = (cy + bh / 2.0 - pad_y) / scale
            x1 = int(clamp(round(x1), 0, w - 1))
            y1 = int(clamp(round(y1), 0, h - 1))
            x2 = int(clamp(round(x2), 0, w - 1))
            y2 = int(clamp(round(y2), 0, h - 1))
            if x2 <= x1 or y2 <= y1:
                continue
            boxes.append([x1, y1, x2 - x1, y2 - y1])
            scores.append(conf)
        if not boxes:
            return []
        idxs = cv2.dnn.NMSBoxes(boxes, scores, self.conf_thresh, self.nms_thresh)
        if len(idxs) == 0:
            return []
        idxs = np.array(idxs).reshape(-1).tolist()[:self.max_det]
        out = []
        for i in idxs:
            x, y, bw, bh = boxes[i]
            out.append((x, y, x + bw, y + bh, scores[i]))
        return out

    def infer(self, bgr):
        blob, scale, pad_x, pad_y = self.preprocess(bgr)
        t0 = time.time()
        output = self.session.run(None, {self.input_name: blob})[0]
        dt = max(time.time() - t0, 1e-6)
        self.last_fps = 1.0 / dt
        return self.postprocess(output, bgr.shape, scale, pad_x, pad_y)

    def draw(self, bgr, boxes):
        out = bgr.copy()
        for x1, y1, x2, y2, conf in boxes:
            cv2.rectangle(out, (x1, y1), (x2, y2), (0, 255, 0), 2)
            label = "obstacle %.2f" % conf
            ty = max(20, y1 - 8)
            cv2.putText(out, label, (x1, ty), cv2.FONT_HERSHEY_SIMPLEX, 0.55, (0, 255, 0), 2, cv2.LINE_AA)
        status = "obstacles: %d" % len(boxes)
        if self.draw_fps:
            status += "  infer: %.1f fps" % self.last_fps
        cv2.putText(out, status, (10, 26), cv2.FONT_HERSHEY_SIMPLEX, 0.7, (0, 255, 255), 2, cv2.LINE_AA)
        return out

    def on_image(self, msg):
        self.frame_idx += 1
        try:
            bgr = self.bridge.imgmsg_to_cv2(msg, desired_encoding="bgr8")
        except CvBridgeError as exc:
            rospy.logwarn("yolo_obstacle_detector cv_bridge: %s", exc)
            return

        if self.output_width > 0 and bgr.shape[1] > self.output_width:
            out_h = int(round(float(bgr.shape[0]) * float(self.output_width) / float(bgr.shape[1])))
            bgr = cv2.resize(bgr, (self.output_width, out_h), interpolation=cv2.INTER_AREA)

        if self.frame_idx % self.process_every_n != 0:
            return
        try:
            self.last_boxes = self.infer(bgr)
        except Exception as exc:
            rospy.logwarn_throttle(2.0, "yolo_obstacle_detector infer failed: %s", exc)
            self.last_boxes = []
        dbg = self.draw(bgr, self.last_boxes)
        try:
            out = self.bridge.cv2_to_imgmsg(dbg, encoding="bgr8")
            out.header = msg.header
            self.pub.publish(out)
        except CvBridgeError as exc:
            rospy.logwarn("yolo_obstacle_detector publish: %s", exc)


def main():
    rospy.init_node("yolo_obstacle_detector", anonymous=False)
    try:
        YoloObstacleDetector()
        rospy.spin()
    except rospy.ROSInterruptException:
        pass
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: ros/catkin_ws/src/simple_robotic_arm/src/yolo_obstacle_detector.cpp
/*
|--------------------------------------------------------------------------
| YOLO ONNX 障碍物检测节点
|--------------------------------------------------------------------------
| 使用 OpenCV DNN 在 C++ 中加载 YOLO ONNX 模型，将检测框绘制到 ROS 调试图像。
|--------------------------------------------------------------------------
*/

#include <algorithm>
#include <chrono>
#include <cmath>
#include <string>
#include <vector>

#include <cv_bridge/cv_bridge.h>
#include <opencv2/dnn.hpp>
#include <opencv2/imgproc.hpp>
#include <ros/package.h>
#include <ros/ros.h>
#include <sensor_msgs/Image.h>

namespace simple_robotic_arm {

/*
|--------------------------------------------------------------------------
| 检测框结构
|--------------------------------------------------------------------------
| 保存 YOLO 后处理后的矩形坐标和置信度。
|--------------------------------------------------------------------------
*/
struct DetectionBox {
  cv::Rect box;
  float confidence = 0.0F;
};

/*
|--------------------------------------------------------------------------
| 图像预处理工具
|--------------------------------------------------------------------------
| 将任意尺寸图像等比例缩放到方形输入，记录缩放比例和填充量供还原坐标。
|--------------------------------------------------------------------------
*/
cv::Mat letterbox(const cv::Mat& bgr, int size, double& scale, int& pad_x, int& pad_y) {
  scale = std::min(static_cast<double>(size) / bgr.cols, static_cast<double>(size) / bgr.rows);
  const int new_w = static_cast<int>(std::round(bgr.cols * scale));
  const int new_h = static_cast<int>(std::round(bgr.rows * scale));
  cv::Mat resized;
  cv::resize(bgr, resized, cv::Size(new_w, new_h));
  cv::Mat canvas(size, size, CV_8UC3, cv::Scalar(114, 114, 114));
  pad_x = (size - new_w) / 2;
  pad_y = (size - new_h) / 2;
  resized.copyTo(canvas(cv::Rect(pad_x, pad_y, new_w, new_h)));
  return canvas;
}

/*
|--------------------------------------------------------------------------
| YOLO 检测节点
|--------------------------------------------------------------------------
| 订阅摄像头图像，按间隔执行 ONNX 推理，发布带框调试图。
|--------------------------------------------------------------------------
*/
class YoloObstacleDetectorNode {
 public:
  YoloObstacleDetectorNode() : private_nh_("~") {
    loadParams();
    net_ = cv::dnn::readNetFromONNX(model_path_);
    net_.setPreferableBackend(cv::dnn::DNN_BACKEND_OPENCV);
    net_.setPreferableTarget(cv::dnn::DNN_TARGET_CPU);
    debug_pub_ = nh_.advertise<sensor_msgs::Image>(debug_topic_, 1);
    image_sub_ = nh_.subscribe(image_topic_, 1, &YoloObstacleDetectorNode::handleImage, this);
    ROS_INFO("yolo_obstacle_detector_cpp: image=%s debug=%s model=%s", image_topic_.c_str(), debug_topic_.c_str(), model_path_.c_str());
  }

 private:
  void loadParams() {
    const std::string package_path = ros::package::getPath("simple_robotic_arm");
    const std::string default_model = package_path + "/models/obstacle_yolo11n.onnx";
    private_nh_.param<std::string>("image_topic", image_topic_, "/usb_cam/image_raw");
    private_nh_.param<std::string>("debug_image_topic", debug_topic_, "/obstacle_detector/debug");
    private_nh_.param<std::string>("model_path", model_path_, default_model);
    private_nh_.param("input_size", input_size_, 512);
    private_nh_.param("conf_thresh", conf_thresh_, 0.05);
    private_nh_.param("nms_thresh", nms_thresh_, 0.45);
    private_nh_.param("max_det", max_det_, 30);
    private_nh_.param("output_width", output_width_, 320);
    private_nh_.param("process_every_n", process_every_n_, 1);
    private_nh_.param("draw_fps", draw_fps_, true);
    process_every_n_ = std::max(1, process_every_n_);
  }

  void handleImage(const sensor_msgs::Image::ConstPtr& msg) {
    frame_index_ += 1;
    if ((frame_index_ % process_every_n_) != 0) {
      return;
    }

    cv_bridge::CvImageConstPtr cv_ptr;
    try {
      cv_ptr = cv_bridge::toCvShare(msg, "bgr8");
    } catch (const cv_bridge::Exception& ex) {
      ROS_WARN("yolo_obstacle_detector cv_bridge: %s", ex.what());
      return;
    }

    cv::Mat bgr = cv_ptr->image.clone();
    if (output_width_ > 0 && bgr.cols > output_width_) {
      const int out_h = static_cast<int>(std::round(static_cast<double>(bgr.rows) * output_width_ / bgr.cols));
      cv::resize(bgr, bgr, cv::Size(output_width_, out_h), 0, 0, cv::INTER_AREA);
    }

    const auto begin = std::chrono::steady_clock::now();
    std::vector<DetectionBox> boxes = infer(bgr);
    const auto elapsed = std::chrono::duration<double>(std::chrono::steady_clock::now() - begin).count();
    last_fps_ = elapsed > 1e-6 ? 1.0 / elapsed : 0.0;
    publishDebug(msg->header, bgr, boxes);
  }

  std::vector<DetectionBox> infer(const cv::Mat& bgr) {
    double scale = 1.0;
    int pad_x = 0;
    int pad_y = 0;
    cv::Mat padded = letterbox(bgr, input_size_, scale, pad_x, pad_y);
    cv::Mat blob = cv::dnn::blobFromImage(padded, 1.0 / 255.0, cv::Size(input_size_, input_size_), cv::Scalar(), true, false);
    net_.setInput(blob);
    cv::Mat output = net_.forward();
    return postprocess(output, bgr.size(), scale, pad_x, pad_y);
  }

  std::vector<DetectionBox> postprocess(const cv::Mat& output, cv::Size image_size, double scale, int pad_x, int pad_y) const {
    cv::Mat rows = output;
    if (output.dims == 3) {
      const int rows_count = output.size[1];
      const int cols_count = output.size[2];
      rows = cv::Mat(rows_count, cols_count, CV_32F, const_cast<float*>(reinterpret_cast<const float*>(output.ptr<float>())));
      if (rows_count == 5 || rows_count < cols_count) {
        rows = rows.t();
      }
    }

    std::vector<cv::Rect> candidate_boxes;
    std::vector<float> scores;
    for (int i = 0; i < rows.rows; ++i) {
      const float* row = rows.ptr<float>(i);
      const float conf = row[4];
      if (conf < conf_thresh_) {
        continue;
      }
      const float cx = row[0];
      const float cy = row[1];
      const float bw = row[2];
      const float bh = row[3];
      int x1 = static_cast<int>(std::round((cx - bw * 0.5F - pad_x) / scale));
      int y1 = static_cast<int>(std::round((cy - bh * 0.5F - pad_y) / scale));
      int x2 = static_cast<int>(std::round((cx + bw * 0.5F - pad_x) / scale));
      int y2 = static_cast<int>(std::round((cy + bh * 0.5F - pad_y) / scale));
      x1 = std::max(0, std::min(x1, image_size.width - 1));
      y1 = std::max(0, std::min(y1, image_size.height - 1));
      x2 = std::max(0, std::min(x2, image_size.width - 1));
      y2 = std::max(0, std::min(y2, image_size.height - 1));
      if (x2 <= x1 || y2 <= y1) {
        continue;
      }
      candidate_boxes.emplace_back(cv::Rect(cv::Point(x1, y1), cv::Point(x2, y2)));
      scores.push_back(conf);
    }

    std::vector<int> keep;
    cv::dnn::NMSBoxes(candidate_boxes, scores, conf_thresh_, nms_thresh_, keep);
    std::vector<DetectionBox> detections;
    for (int index : keep) {
      if (static_cast<int>(detections.size()) >= max_det_) {
        break;
      }
      detections.push_back({candidate_boxes[index], scores[index]});
    }
    return detections;
  }

  void publishDebug(const std_msgs::Header& header, cv::Mat& bgr, const std::vector<DetectionBox>& boxes) {
    for (const auto& detection : boxes) {
      cv::rectangle(bgr, detection.box, cv::Scalar(0, 255, 0), 2);
      cv::putText(bgr, "obstacle " + std::to_string(detection.confidence).substr(0, 4), detection.box.tl() + cv::Point(0, -6), cv::FONT_HERSHEY_SIMPLEX, 0.55, cv::Scalar(0, 255, 0), 2);
    }
    std::string status = "obstacles: " + std::to_string(boxes.size());
    if (draw_fps_) {
      status += " infer: " + std::to_string(last_fps_).substr(0, 4) + " fps";
    }
    cv::putText(bgr, status, cv::Point(10, 26), cv::FONT_HERSHEY_SIMPLEX, 0.7, cv::Scalar(0, 255, 255), 2);
    cv_bridge::CvImage out(header, "bgr8", bgr);
    debug_pub_.publish(out.toImageMsg());
  }

  ros::NodeHandle nh_;
  ros::NodeHandle private_nh_;
  ros::Publisher debug_pub_;
  ros::Subscriber image_sub_;
  cv::dnn::Net net_;
  std::string image_topic_;
  std::string debug_topic_;
  std::string model_path_;
  int input_size_ = 512;
  double conf_thresh_ = 0.05;
  double nms_thresh_ = 0.45;
  int max_det_ = 30;
  int output_width_ = 320;
  int process_every_n_ = 1;
  bool draw_fps_ = true;
  int frame_index_ = 0;
  double last_fps_ = 0.0;
};

}  // namespace simple_robotic_arm

int main(int argc, char** argv) {
  ros::init(argc, argv, "yolo_obstacle_detector");
  simple_robotic_arm::YoloObstacleDetectorNode node;
  ros::spin();
  return 0;
}
